### PR TITLE
CPFP

### PR DESCRIPTION
**Pull request by ptschip**

### CODE REVIEW

File: src/globals.cpp [hunk 1]
```diff
@@ -274,6 +275,9 @@ CTweakRef<bool> displayArchInSubver("net.displayArchInSubver",
     "Show box architecture, 32/64bit, in node user agent string (subver)",
     &fDisplayArchInSubver);
 
+CTweak<bool> miningCPFP("mining.cpfp",
```
Comment (gandrewstone) [inline src/globals.cpp]: let's spell out mining.cpfp as mining.childPaysForParent, but variable name miningCPFP is fine.

File: src/miner.cpp [hunk 2]
```diff
@@ -30,10 +30,16 @@
 #include "validation/validation.h"
 #include "validationinterface.h"
 
+#include <algorithm>
+#include <boost/thread.hpp>
 #include <boost/tuple/tuple.hpp>
 #include <queue>
 #include <thread>
 
+// Track timing information for Score and Pacakge mining.
```
Comment (gandrewstone) [inline src/miner.cpp]: spelling: Package

Also, I'm just going from top to bottom here so may be missing something but shouldn't these be members of BlockAssembler?

File: src/miner.h [hunk 3]
```diff
@@ -22,20 +25,50 @@ class CWallet;
...
+ *  It simply compares the internal memory address of the CTxMemPoolEntry object
+ *  pointed to. This means it has no meaning, and is only useful for using them
+ *  as key in other indexes.
+ */
+struct CompareCTxMemPoolIter
+{
+    bool operator()(const CTxMemPool::txiter &a, const CTxMemPool::txiter &b) const { return &(*a) < &(*b); }
+};
+
+// A comparator that sorts transactions based on number of ancestors.
```
Comment (gandrewstone) [inline src/miner.h]: use doxygen compatible comment format /**

File: src/miner.cpp [hunk 4]
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+        // Test if package fits in the block
+        if (nBlockSize + packageSize > nBlockMaxSize)
+        {
+            if (nBlockSize > nBlockMaxSize * .50)
+            {
+                nPackageFailures++;
+            }
+
+            // If we keep failing then the block must be almost full so bail out here.
+            if (nPackageFailures >= 5)
```
Comment (gandrewstone) [inline src/miner.cpp]: Let's make 5 a symbolic constant, I can imagine that we might want to tune this.

**Comment by ptschip** [comment]
ok done

**Comment by ptschip** [comment]
fixed spelling

as for being part of block assembler, i don't think it makes
sense...these are global vars used to track peformance in the
miner_tests.cpp peformancetest section.

**Comment by ptschip** [comment]
done

File: src/miner.cpp [hunk 5]
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+                return;
+            else
+                continue;
+        }
+
+        // Test the package
```
Comment (gandrewstone) [inline src/miner.cpp]: not a good comment, need to answer why, and try to use different terms than the function name. Maybe "Check the block to determine if this package of transactions will fit within size and sigops limits".

**Comment by ptschip** [comment]
done

**Comment by ptschip** [comment]
fixed, changed this to TestPackageSigOps()

File: src/miner.cpp [hunk 6]
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+    sortedEntries.clear();
+    sortedEntries.insert(sortedEntries.begin(), package.begin(), package.end());
+    std::sort(sortedEntries.begin(), sortedEntries.end(), CompareTxIterByAncestorCount());
+}
+
+// This transaction selection algorithm orders the mempool based
+// on feerate of a transaction including all unconfirmed ancestors.
+//
+// This is accomplished by considering a group of ancestors as a single transaction. We can call these
+// transations, Ancestor Grouped Transactions (AGT). This approach to grouping allows us to process
```
Comment (gandrewstone) [inline src/miner.cpp]: transations -> transactions

File: src/miner.cpp [hunk 7]
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+// packages orders of magnitude faster than other methods of package mining since we no longer have
+// to continuously update the descendant state as we mine part of an unconfirmed chain.
+//
+// There is a theorical flaw in this approach which could happen when a block is almost full. We
```
Comment (gandrewstone) [inline src/miner.cpp]: theorical -> theoretical

File: src/miner.cpp (hunk 5 above)
Comment (ptschip) [reply src/miner.cpp]: done

File: src/miner.cpp (hunk 4 above)
Comment (ptschip) [reply src/miner.cpp]: done

File: src/miner.h (hunk 3 above)
Comment (ptschip) [reply src/miner.h]: done

File: src/miner.cpp (hunk 2 above)
Comment (ptschip) [reply src/miner.cpp]: fixed spelling

as for being part of block assembler, i don't think it makes
sense...these are global vars used to track peformance in the
miner_tests.cpp peformancetest section.

File: src/globals.cpp (hunk 1 above)
Comment (ptschip) [reply src/globals.cpp]: done

File: src/miner.cpp (hunk 6 above)
Comment (ptschip) [reply src/miner.cpp]: done

File: src/miner.cpp (hunk 7 above)
Comment (ptschip) [reply src/miner.cpp]: done

File: src/test/miner_tests.cpp [hunk 8]
```diff
@@ -81,6 +83,297 @@ bool TestSequenceLocks(const CTransaction &tx, int flags)
...
+    uint256 hashHighFeeTx2 = tx.GetHash();
+    mempool.addUnchecked(tx.GetHash(), entry.Fee(10000).FromTx(tx));
+    pblocktemplate = BlockAssembler(chainparams).CreateNewBlock(scriptPubKey);
+    // hashHighFeeTx2 now makes hashFreeTx2 mineable.
+    BOOST_CHECK(pblocktemplate->block.vtx[4]->GetHash() == hashFreeTx2);
+    BOOST_CHECK(pblocktemplate->block.vtx[5]->GetHash() == hashHighFeeTx2);
+    BOOST_CHECK(pblocktemplate->block.vtx[8]->GetHash() == hashLowFeeTx2);
+
+    // Test CPFP with AGT (ancestor grouped transactions)
+    // Add another 0 fee tx to higher fe tx chain. This should also get mined
```
Comment (gandrewstone) [inline src/test/miner_tests.cpp]: fe -> fee

File: src/test/miner_tests.cpp [hunk 9]
```diff
@@ -81,6 +83,297 @@ bool TestSequenceLocks(const CTransaction &tx, int flags)
...
+    // because the total package fees will still be above the minrelaytxfe
```
Comment (gandrewstone) [inline src/test/miner_tests.cpp]: minrelaytxfe -> minrelaytxfee

File: src/main.h [hunk 10]
```diff
@@ -75,13 +75,13 @@ static const CAmount HIGH_MAX_TX_FEE = 100 * HIGH_TX_FEE_PER_KB;
  */
 static const unsigned int DEFAULT_MAX_ORPHAN_TRANSACTIONS = 1000000;
 /** Default for -limitancestorcount, max number of in-mempool ancestors */
-static const unsigned int DEFAULT_ANCESTOR_LIMIT = 25;
+static const unsigned int DEFAULT_ANCESTOR_LIMIT = 500;
```
Comment (gandrewstone) [inline src/main.h]: I think we need to keep these at the BCH default, and then in a separate PR we need to implement XVERSION communication of the setting and preemptive forwarding of tx that would have been rejected from other nodes.

File: src/miner.cpp [hunk 11]
```diff
@@ -30,10 +30,19 @@
...
 #include "validationinterface.h"
 
+#include <algorithm>
+#include <boost/thread.hpp>
 #include <boost/tuple/tuple.hpp>
 #include <queue>
 #include <thread>
 
+// Track timing information for Score and Package mining.
+int64_t nTotalPackage = 0;
```
Comment (gandrewstone) [inline src/miner.cpp]: comment what critsec protects these globals... 
EDIT: actually I've looked closely at these now and they are only used in miner_tests GenerateBlocks which is single threaded.  So they could be garbage in the multi-threaded code and we would not care... but block creation is covered by cs_main anyway -- we only want to make 1 at a time.  So cs_main is protecting them

Comment (ptschip) [reply src/miner.cpp]: i made them atomic anyway

File: src/main.h (hunk 10 above)
Comment (ptschip) [reply src/main.h]: reverted the commit

File: src/test/miner_tests.cpp (hunk 9 above)
Comment (ptschip) [reply src/test/miner_tests.cpp]: done

File: src/test/miner_tests.cpp (hunk 8 above)
Comment (ptschip) [reply src/test/miner_tests.cpp]: done

File: src/miner.cpp [hunk 12]
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
     }
 }
 
+// Skip entries in mapTx that are already in a block.
+bool BlockAssembler::SkipMapTxEntry(CTxMemPool::txiter it)
+{
+    assert(it != mempool.mapTx.end());
+    if (inBlock.count(it))
+        return true;
+    return false;
```
Comment (dagurval) [inline src/miner.cpp]: nit: could just do `return inBlock.count(it)`

File: src/miner.cpp [hunk 13]
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+}
+
+void BlockAssembler::SortForBlock(const CTxMemPool::setEntries &package, std::vector<CTxMemPool::txiter> &sortedEntries)
```
Comment (dagurval) [inline src/miner.cpp:520]: TTOR doesn't exist any more and I don't think the extra code complexity, and the extra test requirements for supporting it is worth it. Thus I think this method should be removed.

If TTOR support is really needed for some reason, then it can be added as a final sorting step (its not used in production, so  the performance hit of sorting is OK)

File: src/miner.cpp [hunk 14]
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
     }
 }
 
+// Skip entries in mapTx that are already in a block.
+bool BlockAssembler::SkipMapTxEntry(CTxMemPool::txiter it)
```
Comment (dagurval) [inline src/miner.cpp]: nit: This method doesn't skip, maybe rename to `IsTxInBlock` or `ShouldSkipTx` ?

File: src/miner.cpp [hunk 15]
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+// to continuously update the descendant state as we mine part of an unconfirmed chain.
+//
+// There is a theoretical flaw in this approach which could happen when a block is almost full. We
+// could for instance end up including a lower fee transaction as part of an ancestor group when
+// in fact it would be better, in terms of fees, to include some other single transaction. This
+// would result in slightly less fees (perhaps a few hundred satoshis) rewarded to the miner. However,
+// this situation is not likely to be seen for two reasons. One, long unconfirmed chains are typically
+// having transactions with all the same fees and Two, the typical child pays for parent scenario has only
+// two transactions with the child having the higher fee. And neither of these two types of packages could
+// cause any loss of fees with this mining algorithm, when the block is nearly full.
```
Comment (dagurval) [inline src/miner.cpp:546]: Fair trade-off. More value in good performance over catching every satoshi.

File: src/miner.cpp [hunk 16]
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
...
+        {
+            continue;
+        }
+        // Test if all tx's are Final
+        if (!TestPackageFinality(ancestors))
+        {
+            continue;
+        }
+
+        // The Package can now be added to the block.
```
Comment (dagurval) [inline src/miner.cpp:621]: We should just assume canonical (see comment above)

File: src/txmempool.h [hunk 17]
```diff
@@ -298,7 +298,7 @@ class CompareTxMemPoolEntryByAncestorFee
 
         if (f1 == f2)
         {
-            return a.GetTx().GetHash() < b.GetTx().GetHash();
+            return a.GetTime() < b.GetTime();
```
Comment (dagurval) [inline src/txmempool.h:301]: Why? Is this related to CPFP?

File: src/miner.cpp [hunk 18]
```diff
@@ -256,6 +256,30 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript &sc
             }
         }
 
+        std::vector<const CTxMemPoolEntry *> vtxe;
+        addPriorityTxs(&vtxe);
+
+        // Mine by package (CPFP) or by score. Score mining is done following this so that we can fill any
```
Comment (dagurval) [inline src/miner.cpp]: > Score mining is done following this

Is it? Where?

**Comment by dagurval** [comment]
Is it possible that ancestor package is 3MB? What happens then if `maxblocksize` is 2MB?

File: src/miner.cpp (hunk 14 above)
Comment (ptschip) [reply src/miner.cpp]: actually i just took this out and is not needed anymore..this was a leftover from Core's work where they did multiple checks here...i just replace skiptmapentry() with inBlock.count(iter)

File: src/miner.cpp (hunk 12 above)
Comment (ptschip) [reply src/miner.cpp]: done, don't need this function anymore.

File: src/miner.cpp (hunk 15 above)
Comment (ptschip) [reply src/miner.cpp:546]: yes

File: src/miner.cpp (hunk 18 above)
Comment (ptschip) [reply src/miner.cpp]: thanks, was incorrect and an older comment.  Removed the last part.

File: src/txmempool.h (hunk 17 above)
Comment (ptschip) [reply src/txmempool.h:301]: Yes, gives a perfect fill in the block when for instance we have a long package which is greater than the maxblocksize (or the remaining amount of block to be filled) and all the fees for each transaction are equal.  This sorting does two things.  This way we end up finding the part of the package that first perfectly into the block, because we don't bail out early due to pacakge insertion failures, while also preserving some sense of fairness that the first transaction that arrived in the mempool has priority over the one that followed.  (Remember in BU and also ABC there is no longer any random delay in forwarding inv's anymore.)

**Comment by ptschip** [comment]
> Is it possible that ancestor package is 3MB? What happens then if `maxblocksize` is 2MB?

Very good question. Partly answered above with the  time sorting of the txns but what if the txns all have different fees?  Then the time sorting doesn't work and we have a potential problem because we can easily get 5 package failures and end up bailing out of building the block and end up with an empty block!  The solution  here is not to start counting package failures until we have partially filled the block.  If you look further down where we test if the package fits the block, you'll see we don't start counting package failures until we're > 50% full.

File: src/miner.cpp (hunk 16 above)
Comment (ptschip) [reply src/miner.cpp:621]: Good point but we can't because we still have both in our regression tests such as ctor.py and also in the miner we later either do the sort or not based on the "canonical" flag.  I think if we wan't to remove support for TTOR then it should be done in a different PR.

File: src/miner.cpp (hunk 13 above)
Comment (ptschip) [reply src/miner.cpp:520]: Same comment as below:  "Good point but we can't because we still have both in our regression tests such as ctor.py and also in the miner we later either do the sort or not based on the "canonical" flag. I think if we wan't to remove support for TTOR then it should be done in a different PR."

Comment (gandrewstone) [reply src/miner.cpp:520]: I agree, let's leave TTOR around for a while longer.
